components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/internal/service/impl/UserAccountRecoveryManager.java

```diff
@@ -284,53 +295,99 @@ public String getUsernameByClaims(Map<String, String> claims, String tenantDomai
+                List<ExpressionCondition> expressionConditionList =
+                        getExpressionConditionList(claims, domain, claimManager);
+
+                if (!expressionConditionList.isEmpty()) {
+                    Condition operationalCondition = getOperationalCondition(expressionConditionList);
+                    /* Get the users list that matches with the condition
+                       limit : 2, offset : 1, sortBy : null, sortOrder : null */
+                    resultedUserList.addAll(carbonUM.getUserListWithID(operationalCondition, domain,
+                            UserCoreConstants.DEFAULT_PROFILE, 2, 1, null, null));
                 }
-                // Get the matching user list for the given claim.
-                String[] matchedUserList = getUserList(tenantId, key, value);
-                // Check for duplicates with the already retrieved users, if the matched users for the claim.
-                if (!ArrayUtils.isEmpty(matchedUserList)) {
-                    // In the first iteration resultedUserList will be empty.
-                    if (ArrayUtils.isNotEmpty(resultedUserList)) {
-                        // Remove the users from already matched list who are not in the recently matched list.
-                        resultedUserList = getCommonUserEntries(resultedUserList, matchedUserList, key, value);
-                        if (ArrayUtils.isEmpty(resultedUserList)) {
-                            if (log.isDebugEnabled()) {
-                                log.debug("No user matched for given claims");
-                            }
-                            return StringUtils.EMPTY;
-                        }
-                    } else {
-                        resultedUserList = matchedUserList;
-                    }
-                } else {
+                else {
+                    continue;
+                }
+
+                if (resultedUserList.size() > 1) {
                     if (log.isDebugEnabled()) {
-                        String message = String.format("No users matched for claim : %s for value : %s", key, value);
-                        log.debug(message);
+                        log.debug("Multiple users matched for given claims set : " +
+                                Arrays.toString(resultedUserList.toArray()));
                     }
-                    return StringUtils.EMPTY;
+                    throw Utils.handleClientException(
+                            IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_MULTIPLE_MATCHING_USERS, null);
                 }
             }
-        }
-        // Return matched user.
-        if (ArrayUtils.isNotEmpty(resultedUserList) && resultedUserList.length == 1) {
-            return resultedUserList[0];
-        } else {
+            // Return empty when no users are found.
+            if (resultedUserList.isEmpty()) {
+                return StringUtils.EMPTY;
+            } else {
```

Comment:
We don't need the else block here

Reply:
Resolved by https://github.com/wso2-extensions/identity-governance/pull/623/commits/fa0dbdc33ab038eb8cae2b8d2d5fffd765503470